shell/platform/windows/windows_lifecycle_manager.cc

```diff
@@ -163,4 +183,105 @@ void WindowsLifecycleManager::BeginProcessingClose() {
   process_close_ = true;
 }
 
+void WindowsLifecycleManager::SetLifecycleState(AppLifecycleState state) {
+  if (state_ == state) {
+    return;
+  }
+  state_ = state;
+  if (engine_) {
+    const char* state_name = AppLifecycleStateToString(state);
+    engine_->SendPlatformMessage("flutter/lifecycle",
+                                 reinterpret_cast<const uint8_t*>(state_name),
+                                 strlen(state_name), nullptr, nullptr);
+  }
+}
+
+void WindowsLifecycleManager::OnWindowStateEvent(HWND hwnd,
+                                                 WindowStateEvent event) {
+  // Synthesize an unfocus event when a focused window is hidden.
+  if (event == WindowStateEvent::kHide &&
+      focused_windows_.find(hwnd) != focused_windows_.end()) {
+    OnWindowStateEvent(hwnd, WindowStateEvent::kUnfocus);
+  }
+
+  std::lock_guard guard(state_update_lock_);
+  switch (event) {
+    case WindowStateEvent::kShow: {
+      bool first_shown_window = visible_windows_.empty();
+      auto pair = visible_windows_.insert(hwnd);
+      if (first_shown_window && pair.second &&
+          state_ == AppLifecycleState::kHidden) {
+        SetLifecycleState(AppLifecycleState::kInactive);
+      }
+      break;
+    }
+    case WindowStateEvent::kHide: {
+      bool present = visible_windows_.erase(hwnd);
+      bool empty = visible_windows_.empty();
+      if (present && empty &&
+          (state_ == AppLifecycleState::kResumed ||
+           state_ == AppLifecycleState::kInactive)) {
+        SetLifecycleState(AppLifecycleState::kHidden);
+      }
+      break;
+    }
+    case WindowStateEvent::kFocus: {
+      bool first_focused_window = focused_windows_.empty();
+      auto pair = focused_windows_.insert(hwnd);
+      if (first_focused_window && pair.second &&
+          state_ == AppLifecycleState::kInactive) {
+        SetLifecycleState(AppLifecycleState::kResumed);
+      }
+      break;
+    }
+    case WindowStateEvent::kUnfocus: {
+      if (focused_windows_.erase(hwnd) && focused_windows_.empty() &&
+          state_ == AppLifecycleState::kResumed) {
+        SetLifecycleState(AppLifecycleState::kInactive);
+      }
+      break;
+    }
+  }
+}
+
+bool WindowsLifecycleManager::ExternalWindowMessage(HWND hwnd,
+                                                    UINT message,
+                                                    WPARAM wparam,
+                                                    LPARAM lparam) {
+  std::optional<flutter::WindowStateEvent> event = std::nullopt;
+
+  switch (message) {
```

Comment:
Please add a TODO for the `WM_CLOSE` message here with a link to a tracking issue.